ulib/experimental/Steel.SelEffect.Common.fsti

```diff
@@ -216,6 +216,88 @@ inline_for_extraction noextract let rm : CE.cm vprop req =
      star_commutative
      star_congruence
 
+val vrefine_hp (v: vprop) (p: (normal (t_of v) -> Tot prop)) : Tot (slprop u#1)
+
+val interp_vrefine_hp (v: vprop) (p: (normal (t_of v) -> Tot prop)) (m: mem) : Lemma
+  (interp (vrefine_hp v p) m <==> (interp (hp_of v) m /\ p (sel_of v m)))
+//  [SMTPat (interp (vrefine_hp v p) m)] // FIXME: this pattern causes Z3 "wrong number of argument" errors
+
+[@__steel_reduce__]
+let vrefine_t (v: vprop) (p: (normal (t_of v) -> Tot prop)) : Tot Type
+= (x: t_of v {p x})
+
+val vrefine_sel (v: vprop) (p: (normal (t_of v) -> Tot prop)) : Tot (selector (vrefine_t v p) (vrefine_hp v p))
+
+val vrefine_sel_eq (v: vprop) (p: (normal (t_of v) -> Tot prop)) (m: hmem (vrefine_hp v p)) : Lemma
+  (
+    interp (hp_of v) m /\
+    vrefine_sel v p m == sel_of v m
+  )
+//  [SMTPat ((vrefine_sel v p) m)] // FIXME: this pattern causes Z3 "wrong number of argument" errors
+
+[@__steel_reduce__]
+let vrefine' (v: vprop) (p: (normal (t_of v) -> Tot prop)) : Tot vprop' = {
+  hp = vrefine_hp v p;
+  t = vrefine_t v p;
+  sel = vrefine_sel v p;
+}
+
+[@__steel_reduce__]
+let vrefine (v: vprop) (p: (normal (t_of v) -> Tot prop)) = VUnit (vrefine' v p)
+
+val vdep_hp (v: vprop) (p: ( (t_of v) -> Tot vprop)) : Tot (slprop u#1)
+
+val interp_vdep_hp (v: vprop) (p: ( (t_of v) -> Tot vprop)) (m: mem) : Lemma
+  (interp (vdep_hp v p) m <==> (interp (hp_of v) m /\ interp (hp_of v `Mem.star` hp_of (p (sel_of v m))) m))
+
+[@__steel_reduce__]
+let vdep_payload
+  (v: vprop) (p: ( (t_of v) -> Tot vprop))
+  (x: t_of v)
+: Tot Type
+= t_of (p x)
+
+[@__steel_reduce__]
+let vdep_t (v: vprop) (p: ( (t_of v) -> Tot vprop)) : Tot Type
+= dtuple2 (t_of v) (vdep_payload v p)
```

Comment:
How does this dtuple2 interact with normalization?

Reply:
I haven't investigated interaction with normalization yet. I initially tried to insert `normal` here and there but they went in the way of proving `intro_vdep`, `elim_vdep`.
By the way I don't intend to rely on normalization for the dependent part `p`. More generally, I have yet to understand the role played by the various `normal` annotations on other combinators. So if you believe that users could benefit from it, I'd propose to add separate combinators with `normal` annotations, calling into those without.